src/Symfony/Component/Config/Definition/EnumNode.php

```diff
@@ -78,13 +88,36 @@ protected function finalizeValue(mixed $value): mixed
     {
         $value = parent::finalizeValue($value);
 
-        if (!\in_array($value, $this->values, true)) {
-            $ex = new InvalidConfigurationException(\sprintf('The value %s is not allowed for path "%s". Permissible values: %s', json_encode($value), $this->getPath(), $this->getPermissibleValues(', ')));
-            $ex->setPath($this->getPath());
+        if ($this->enumFqcn) {
+            if (\is_a($value, $this->enumFqcn, true)) {
+                return $value;
+            }
+
+            if (\is_string($value) || \is_int($value)) {
+                try {
+                    $value = $this->enumFqcn::from($value);
+                } catch (\ValueError) {
+                    throw $this->createInvalidValueException($value);
+                }
+            } elseif ($value instanceof \UnitEnum) {
+                throw new InvalidConfigurationException(sprintf('The value should be part of the "%s" enum, got a value from the "%s" enum.', $this->enumFqcn, $value::class));
+            } else {
+                throw new InvalidConfigurationException(sprintf('The value should either be an integer, a string or part of the "%s" enum, got "%s".', $this->enumFqcn, get_debug_type($value)));
+            }
```

Comment:
Just thinking that right now if you pass something that is not int|string, it will tell you you should pass an int or string, or a value of the enum. But IMO it'd be better to just list the possible values using createInvalidValueException, because in most cases int will not be valid for example.

Just an idea:

```suggestion
            if ($value instanceof \UnitEnum) {
                throw new InvalidConfigurationException(sprintf('The value should be part of the "%s" enum, got a value from the "%s" enum.', $this->enumFqcn, $value::class));
            }
            if (\is_string($value) || \is_int($value)) {
                $case = $this->enumFqcn::tryFrom($value);
                if (null !== $case) {
                    return $case;
                }
            }

            throw $this->createInvalidValueException($value);
```

I'd maybe edit the error message in `createInvalidValueException` to also mention the enum class though for completeness.

Reply:
I like it :) I also tried a new error message that includes the enum class